New tool addition: Sourmash tool suite

Smkingsize opened the pull request:
FOR CONTRIBUTOR:
* [x] I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] License permits unrestricted use (educational + commercial)
* [x] This PR adds a new tool or tool collection
* [ ] This PR updates an existing tool or tool collection
* [ ] This PR does something else (explain below)

Smkingsize committed:
lets start here

Smkingsize committed:
sourmash sketch added parameters

Smkingsize committed:
soursketch almost ready

Smkingsize committed:
sourmash better specific tests

Smkingsize committed:
added new tools scelet files compare and compute

Smkingsize committed:
sketch is ready

Smkingsize committed:
fixed outputs of sourmash plot

Smkingsize committed:
Merge branch 'galaxyproject:main' into sourmash

SaimMomin12 commented on tool_collections/sourmash/sourmash_plot/sourmash_plot.xml:
```diff
@@ -0,0 +1,112 @@
+            <output name="matrix" ftype="png">
+                <assert_contents>
+                    <has_size size="12k" delta="1k"/>
+                </assert_contents>
```
You can use some other asserts here

SaimMomin12 reviewed (commented):
Some comments inline. Would be nice if we use the following structure instead of individual folders
-- sourmash
   -- test-data
   -- sourmash_plot.xml
   -- sourmash_sketch.xml
   -- macros.xml
   -- shed.yml (This needs a suite defination)

Smkingsize committed:
re-structure suit file tree (no sub-directories)
added some macros
added initial .shed.yml file (not final)
added better assers

Smkingsize committed:
added valid categories.

Smkingsize committed:
moved the tool to tools/
removed too big files

Smkingsize committed:
added macros, edited help sections

Smkingsize committed:
fixed picklist macro

Smkingsize committed:
added sourmash index

Smkingsize committed:
tested 90% of index parameters

Smkingsize committed:
fixed failing tests

Smkingsize committed:
wrote better tests for index (not using has_size).
sourmash index is ready!

Smkingsize committed:
added zipped test-files
added sourmash_search

Smkingsize committed:
added params to macros.
fixed 1 test in sourmash_index

Smkingsize committed:
added sourmash prefetch

Smkingsize committed:
add sourmash multigather

Smkingsize committed:
added sourmas tax metagenome

Smkingsize committed:
added some more subcommands of tax with tests

Smkingsize committed:
added grep and sum

Smkingsize committed:
wrapped 'prepare' in a more logicall way. and some other minor updates like help sections and labels.

Smkingsize committed:
Merge branch 'galaxyproject:main' into sourmash

Smkingsize committed:
wrapped grep, cat and describe

Smkingsize committed:
added and tested merge and split

Smkingsize committed:
added inflate and downsample and test data

Smkingsize committed:
added all sourmash signature subcommands expect collect!

Smkingsize committed:
fixed failed tests and wrote a better structured Help section!

Smkingsize committed:
added kmers!
Suite is ready for review!

Smkingsize committed:
re-run github workflows again!

Smkingsize committed:
edited .shed.yml

Smkingsize committed:
fixed .shed.yml.
added sourmash.sig to test linting

Smkingsize committed:
replaced json with sourmash.sig everywhere!

SaimMomin12 commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,167 @@
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="@TOOL_VERSION@">sourmash</requirement>
+            <requirement type="package" version="*">zip</requirement>
```
```suggestion
            <requirement type="package" version="3.0">zip</requirement>
```

SaimMomin12 commented on tools/sourmash/sourmash_compare.xml:
```diff
@@ -0,0 +1,184 @@
+        </test>
+    </tests>
+    <help><![CDATA[
+        **What the tool does?**
```
Can we add a bit more of help from here? https://sourmash.readthedocs.io/en/latest/command-line.html#sourmash-compare-compare-many-signatures

SaimMomin12 commented on tools/sourmash/sourmash_index.xml:
```diff
@@ -0,0 +1,252 @@
+            <when value="build">
+                <param name="index_type" argument="-F" type="select" label="Index Type" help="choose the database index type">
+                    <option value="SBT" selected="true">SBT - Sequence Bloom Trees</option>
+                    <option value="rocksdb" >RocksDB</option>
```
```
  -F, --index-type {SBT,rocksdb,zip}
                        type of index to build (default: SBT)
```

ZIP missing?

SaimMomin12 commented on tools/sourmash/sourmash_multigather.xml:
```diff
@@ -0,0 +1,76 @@
+    <inputs>
+        <param name="queries" type="data" format="sourmash.sig" multiple="true" label="Query signatures" help="One or more query signature files to gather against databases"/>
+        <param name="databases" type="data" format="zip,sourmash.sig" multiple="true" label="Database(s)" help="One or more signature collections / indexed databases to search"/>
+        <param name="threshold_bp" type="integer" value="50000" label="Threshold (bp)" help="Minimum number of base pairs to report in a match. Default is 50,000 bp"/>
```
```suggestion
        <param aragument="--threshold-bp" type="integer" value="50000" label="Threshold (bp)" help="Minimum number of base pairs to report in a match. Default is 50,000 bp"/>
```

SaimMomin12 commented on tools/sourmash/sourmash_index.xml:
```diff
@@ -0,0 +1,252 @@
+<tool id="sourmash_index" name="sourmash index" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <macros>
```
`<description>` tag missing

SaimMomin12 commented on tools/sourmash/sourmash_prefetch.xml:
```diff
@@ -0,0 +1,195 @@
+        <section name="output_options" title="Outputs" expanded="true">
+            <param name="save_matches" type="boolean" checked="false" label="Save matching signatures" help="Save all matching signatures (as a multi-signature .sig)."/>
+            <param name="save_matching_hashes" type="boolean" checked="false" label="Save matching hashes" help="Save one signature containing hashes that matched any DB signature at/above threshold."/>
+            <param name="save_unmatched_hashes" type="boolean" checked="false" label="Save unmatched hashes" help="Save one signature containing the complement of matching hashes."/>
```
```suggestion
            <param argument="--save-matches" type="boolean" checked="false" label="Save matching signatures" help="Save all matching signatures (as a multi-signature .sig)."/>
            <param argument="--save-matching-hashes" type="boolean" checked="false" label="Save matching hashes" help="Save one signature containing hashes that matched any DB signature at/above threshold."/>
            <param argument="--save-unmatched-hashes" type="boolean" checked="false" label="Save unmatched hashes" help="Save one signature containing the complement of matching hashes."/>
```

SaimMomin12 commented on tools/sourmash/sourmash_search.xml:
```diff
@@ -0,0 +1,181 @@
+                <option value="--max-containment">Max-containment (max directional)</option>
+            </param>
+            <param name="threshold" argument="-t" type="float" value="0.08" min="0.0" max="1.0" label="Similarity threshold" help="Minimum score to report matches"/>
+        </section>
```
```
  --save-matches FILE   output matching signatures to the specified file
  --best-only           report only the best match (with greater speed)
  -n, --num-results N   number of results to display to user; 0 to report all

```
??

SaimMomin12 commented on tools/sourmash/sourmash_sketch.xml:
```diff
@@ -0,0 +1,288 @@
+<tool id="sourmash_sketch" name="sourmash sketch" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <macros>
+        <import>macros.xml</import>
+    </macros>
```
`<description>` tag missing

SaimMomin12 commented on tools/sourmash/sourmash_signature.xml:
```diff
@@ -0,0 +1,729 @@
+        #else if $sig.subcommand == "check":
+            @list_inputs@
+            --picklist
+            $sig.list::manifest
```
??

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,167 @@
+                <option value="false" selected="true">No</option>
+            </param>
+            <when value="true">
+                <param name="picklist_file" type="data" format="csv" label="Picklist file" help="input signatures files (created with souramsh sksetch)"/>
```
Could we remove the conditional by making the input optional. For `pickstyle` we could document in the help that it's only used if `picklist_file` is given?

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,167 @@
+    ]]></token>
+    <xml name="common_params">
+        <section name="sig_select" title="Signature selectors" expanded="false">
+            <param argument="--ksize" type="integer" optional="true" min="1" max="64" label="K-mer size" help="K-mer size to select. Must match the k-mer size of input signatures. Common values: 21, 31, 51"/>
```
There seems to be an overlap with `sig_ksize` macro?

Smkingsize committed:
added description too all tools.
added option zip to sourmash index.
resolved some typos.

Smkingsize commented on tools/sourmash/sourmash_sketch.xml:
```diff
@@ -0,0 +1,288 @@
+        -o $out_file1
+    ]]></command>
+    <inputs>
+        <param name="input" type="data" format="fasta,fastq" label="Input file"/>
```
yes gzipped data is supported

Smkingsize committed:
turned merged cond into optional!
fixed all datatypes.
seperated tax subommands into seperate tools.
picklist cond -> optional.
added missing parameters.
ordered macros and used them where possible!
resolved all other issues!

bernt-matthias commented on tools/sourmash/sourmash_search.xml:
```diff
@@ -0,0 +1,241 @@
+            -o $search_results
+        #end if
+        #if $output_section.output_log:
+            > $log_file
```
single quotes.

Smkingsize committed:
resolved issues!

bernt-matthias commented on tools/sourmash/sourmash_index.xml:
```diff
@@ -0,0 +1,270 @@
+                <param name="index_type" argument="-F" type="select" label="Index Type" help="choose the database index type">
+                    <option value="SBT" selected="true">SBT - Sequence Bloom Trees</option>
+                    <option value="rocksdb" >RocksDB</option>
+                    <option value="zip" >zip</option>
```
I mean the tool docs.

bernt-matthias commented on tools/sourmash/sourmash_prefetch.xml:
```diff
@@ -0,0 +1,199 @@
+    ]]></command>
+    <inputs>
+        <param name="query_sig" type="data" format="sourmash.sig" label="Query signature" help="Scaled (FracMinHash) signature used as the query (often a metagenome or genome signature)."/>
+        <param name="databases" type="data" format="zip,sourmash.sig" multiple="true" label="Database(s)" help="One or more signature collections / indexed databases to search."/>
```
Is zip always referring to multiple sig files in one zip? 

What is the advantage of zip? 

Sorry about all the questions.

SaimMomin12 commented:
@bernt-matthias I was wondering if this PR is good for a merge? Some of our users would like to use this tool.

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,182 @@
+    #end for
+    ]]></token>
+    <!-- tool specific: sourmash_signature.xml-->
+    <token name="@signatures_command_sig@"><![CDATA[
```
This seems unused

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,182 @@
+    <token name="@Picklist_commandline@"><![CDATA[
+    #if $picklist.picklist_file: 
+        #set exp = 'list.csv:md5:md5:%s' % (str($picklist.pickstyle))
+        --picklist $exp
```
```suggestion
        --picklist 'list.csv:md5:md5:$picklist.pickstyle'
```

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,182 @@
+        <param name="signatures" type="data" format="sourmash.sig" multiple="true" label="signature files" help="Input signature files."/>
+    </xml>
+    <!-- generic -->
+    <token name="@signatures_command@"><![CDATA[
```
We have the (unwritten?) convention to use all upper case for macro token names.

bernt-matthias commented on tools/sourmash/sourmash_compare.xml:
```diff
@@ -0,0 +1,191 @@
+        @Picklist_copy_file@ 
+        
+        sourmash compare
+        @Picklist_commandline@
```
Just a minor one. It would be easier to review if the order of the parameters in the inputs section is a close as possible to the order of their use in the commands block.

bernt-matthias commented on tools/sourmash/sourmash_index.xml:
```diff
@@ -0,0 +1,271 @@
+        ]]></command>
+    <inputs>
+        <conditional name="mode_cond">
+            <param name="mode_param" type="select" label="Mode" help="Chose weather to 'Build' new indexed databse from signatures or 'Append' signatures to existing databse.">
```
typo: databse

bernt-matthias commented on tools/sourmash/sourmash_index.xml:
```diff
@@ -0,0 +1,271 @@
+            --append indexed.sbt.zip
+        #end if
+
+        #for $i in $signatures:
```
should `@signatures_command@` be used here?

bernt-matthias commented on tools/sourmash/sourmash_index.xml:
```diff
@@ -0,0 +1,271 @@
+            ${i}
+        #end for
+
+        @common_params_commandline@
```
Would be good / easier to understand the tool if there is a 1:1 relation between the used macro tokens and the corresponding xml macros for the input section.

bernt-matthias commented on tools/sourmash/sourmash_multigather.xml:
```diff
@@ -0,0 +1,81 @@
+        <test expect_num_outputs="3">
+            <param name="query" value="GCF_000005845.2_ASM584v2_genomic.fna.gz.sig"/>
+            <param name="db" value="indexed.sbt.zip"/>
+            <assert_stderr>
```
Please add assertions on the output collections.

bernt-matthias commented on tools/sourmash/sourmash_sketch.xml:
```diff
@@ -0,0 +1,272 @@
+            </when>
+        </conditional>
+        <section name="advanced_settings" title="Advanced Settings" expanded="no">
+            <param name="merge" type="data" optional="true" label="Add more input files" multiple="true" format="fasta,fastqsanger,fastqillumina" help="adding more files would have the effects of merging the sequence records before signature creation"/>
```
Can we add gz? In the worst case gunzip in the wrappers.

SaimMomin12 committed:
Changes as per code review

SaimMomin12 committed:
Some more changes

SaimMomin12 commented:
@bernt-matthias Took a while but the changes suggested by you are implemented. The PR is ready from my side.

bernt-matthias commented on tools/sourmash/sourmash_sketch.xml:
```diff
@@ -0,0 +1,274 @@
+        -o $out_file1
+    ]]></command>
+    <inputs>
+        <param name="input" type="data" format="fasta,fastqsanger,fastqillumina" label="Input file"/>
```
Please add the gz-ipped variants of the datatypes.

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,206 @@
+    ]]></token>
+
+    <!-- tool specific: sourmash_signature.xml-->
+    <token name="@SIGNATURES_COMMAND_SIG@"><![CDATA[
```
This seems unused?

SaimMomin12 replied:
It is used in `sourmash_signature.xml` at multiple places

bernt-matthias commented on tools/sourmash/sourmash_multigather.xml:
```diff
@@ -0,0 +1,126 @@
+            <expand macro="ksize_macro"/>
+            <expand macro="mol_type_macro"/>
+            <expand macro="scaled_macro"/>
+        </section>
```
It should be more obvious that these correspond to the `COMMON_PARAMS_COMMANDLINE`. Maye add a macro?

SaimMomin12 replied:
Done

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,206 @@
+    #end if
+    #if $sig_select.scaled:
+        --scaled $sig_select.scaled
+    #end if
```
Reuse MUL_TYPE_COMMAND, SCALED_COMMAND, and KSIZE_COMMAND

SaimMomin12 replied:
Done

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,206 @@
+
+    <!-- ksize -->
+    <xml name="ksize_macro">
+        <param argument="--ksize" type="integer" optional="true" min="1" max="64" label="K-mer size" help="Select the k-mer size of the signature to use. It must match the k-mer size of one or more of the input signatures. Common values: 21, 31, 51."/>
```
What is used if nothing is selected. If possible make it mandatory and use the default value implemented in the underlying software.

SaimMomin12 replied:
I'd prefer to keep `--ksize` optional. sourmash's own CLI keeps it optional and auto-uses the k-mer size present in the input signatures when it's unambiguous. One only need to set it when the signatures carry more than one k-size. The k=31 default only applies at sketch time - forcing 31 here would break signatures sketched at other k-sizes (e.g. 21/51). I've add this to help text to document exactly what happens when it's left blank.

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,206 @@
+    <!-- Picklist -->
+    <xml name="picklist_macro">
+        <section name="picklist" title="Picklist" expanded="false">
+            <param name="picklist_file" type="data" optional="true" format="csv" label="Picklist file" help="Input signatures files created with sourmash sketch"/>
+            <param name="pickstyle" type="select" optional="true" label="Pick style" help="Whether to include or exclude picked signatures. Used only when picklist file is uploaded!">
+                <option value="include" selected="true">include</option>
+                <option value="exclude">exclude</option>
+            </param>
```
Is it possible to provide an include + exclude file to the underlying software. If so I would change this to tow optional inputs (one for an include and and for the exclude lists file)

SaimMomin12 replied:
sourmash's `--picklist `accepts only a single picklist per command invocation, so supplying an include list and an exclude list at the same time isn't supported by the underlying tool. I've kept the single picklist_file + pickstyle (include/exclude) select, which maps 1:1 to that one `--picklist` argument.

bernt-matthias commented on tools/sourmash/macros.xml:
```diff
@@ -0,0 +1,206 @@
+    </token>
+    
+    <!-- Taxonomy Inputs -->
+    <token name="@TAX_INPUTS_COMMAND@"><![CDATA[
```
Can we have a 1:1 relation between parameters used in the tokens and defined in the macros?

SaimMomin12 replied:
Done

SaimMomin12 committed:
Ammend review comments

SaimMomin12 committed:
Update sourmash_plots with better tests

bernt-matthias commented on tools/sourmash/sourmash_multigather.xml:
```diff
@@ -0,0 +1,122 @@
+<tool id="sourmash_multigather" name="sourmash multigather" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <description>search multiple query signatures against one or more databases or signature collections</description>
```
```suggestion
    <description>search query signatures against signature collections</description>
```


Also define the term database? I guess it's some zip file?

bernt-matthias commented on tools/sourmash/sourmash_multigather.xml:
```diff
@@ -0,0 +1,122 @@
+    ]]></command>
+    <inputs>
+        <param argument="--query" type="data" format="sourmash.sig" multiple="true" label="Query signatures" help="One or more query signature files to gather against databases"/>
+        <param argument="--db" type="data" format="zip,sourmash.sig" multiple="true" label="Database(s)" help="One or more signature collections / indexed databases to search"/>
```
Do we need zip? How would users generate these zip files?

SaimMomin12 replied:
Yes, the `sourmash_index` gives the output database in a ZIP file. This ZIP file can be used in any of the downstream tools.

SaimMomin12 committed:
Update tools/sourmash/sourmash_multigather.xml

Co-authored-by: M Bernt <m.bernt@ufz.de>
